network/ipfs_impl.go

```diff
@@ -112,14 +113,15 @@ func (s *streamMessageSender) Connect(ctx context.Context) (network.Stream, erro
 	}
 
 	s.stream = stream
+	s.connected = true
 	return s.stream, nil
 }
 
 // Reset the stream
 func (s *streamMessageSender) Reset() error {
 	if s.stream != nil {
 		err := s.stream.Reset()
-		s.stream = nil
+		s.connected = false
```

Comment:
Why not just set the stream to nil? That will free up the resources as well.

Reply:
I believe the crash was caused because we were calling SupportsHave() after a Reset():
```
func (s *streamMessageSender) SupportsHave() bool {
	return s.bsnet.SupportsHave(s.stream.Protocol())
}
```

Reply:
Ah, I see. We shouldn't even construct a streamMessageSender till we have the stream.